kernelci/data/kernelci_api.py

```diff
@@ -83,6 +92,44 @@ def get_nodes_by_commit_hash(self, commit_hash):
     def get_node_from_event(self, event):
         return self.get_node(event.data['id'])
 
+    def pubsub_event_filter(self, sub_id, event):
+        """Filter Pub/Sub events
+
+        Filter received Pub/Sub event using provided filter dictionary.
+        Return True if the event matches with the filter, otherwise False.
+        """
+        event_filter_status = True
+        filters = self._filters.get(sub_id)
+        if not filters:
+            return event_filter_status
+        for key, value in filters.items():
+            if key not in event.keys():
+                continue
+            if isinstance(value, dict):
+                for sub_key, sub_value in value.items():
+                    if sub_key not in event.get(key):
+                        continue
+                    if sub_value != event.get(key).get(sub_key):
+                        return
+            elif value != event[key]:
+                return
+
+        return event
+
+    def receive_node(self, sub_id):
+        """
+        Listen to all the events on 'node' channel and apply filter on it.
+        Return node if event matches with the filter.
+        """
+        path = '/'.join(['listen', str(sub_id)])
+        while True:
+            resp = self._get(path)
+            event = from_json(resp.json().get('data'))
+            node = self.get_node_from_event(event)
+            node['op'] = event.data['op']
```

Comment:
I think it's not great practice to change the data in the node dictionary as it's not a node any more. Instead, the node and event objects could be passed to the filter function which would then try to match the filters with both of them. Or you could call the function twice:
```python
if all(self.pubsub_event_filter(sub_id, obj) for obj in [node, event]):
    return node
````
Then the `node` object being returned will be what was received from the API with any modifications (i.e. without adding `op` to it).